pyrk/th_system.py

```diff
@@ -104,3 +107,262 @@ def custom(self, t_b, t_env, res):
+                (component.T[t_idx]*2 - d['t_in']).magnitude,
+                component.T[t_idx].magnitude)
+        for interface, d in component.cond.iteritems():
+            env = self.comp_from_name(interface)
+            Qcond = self.conduction(t_b=component.T[t_idx],
+                                    t_env=env.T[t_idx],
+                                    r_b=d['r_b'],
+                                    r_env=d['r_env'],
+                                    k=d['k'])
+            to_ret -= Qcond/cap
+            assert (Qcond*(component.T[t_idx]-env.T[t_idx])).magnitude >= 0, \
+                'conduction from %s to %s, from temp %f to %f is wrong %f' % (
+                    component.name, env.name, component.T[t_idx].magnitude,
+                    env.T[t_idx].magnitude, Qcond.magnitude)
+        for interface, d in component.conv.iteritems():
+            env = self.comp_from_name(interface)
+            Qconv = self.convection(t_b=component.T[t_idx],
+                                    t_env=env.T[t_idx],
+                                    h=d['h'],
+                                    A=d['area'])
+            to_ret -= Qconv/cap
+            assert (Qconv*(component.T[t_idx]-env.T[t_idx])).magnitude >= 0,\
+                'convection from %s to %s, from temp %f to %f is wrong %f' % (
+                    component.name, env.name, component.T[t_idx].magnitude,
+                    env.T[t_idx].magnitude, Qconv.magnitude)
+        return to_ret
+
+    def conduction(self, t_b, t_env, r_b, r_env, k):
+        """
+        heat transfer by conduction(watts)
+        :param t_b: The temperature of the body
+        :type t_b: float.
+        :param t_env: The temperature of the environment
+        :type t_env: float.
+        :param r_b: The radius of body
+        :type r_b: float.
+        :param r_env: The radius of environment
+        :type r_env: float.
+        :param k: the thermal conductivity
+        :type k: float.
+        """
+        num = 4*math.pi*k*(t_b-t_env)
+        denom = (1/r_b - 1/r_env)
+        return num/denom
+
+    def advection(self, t_out, t_in, m_flow, cp):
+        ''' calculate heat transfer by advection in watts
+        :param t_out: outlet temperature
+        :type t_out: float
+        :param t_in: inlet temperature
+        :type t_in: float
+        :param m_flow: mass flow rate though the control volume
+        :type m_flow: float
+        :param cp: specific heat capacity of the fluid
+        :type cp: float
+        '''
+        if t_out > t_in:
+            return m_flow*cp*(t_out-t_in)
+        else:
+            return 0*units.watts
+
+
+class THSystemSphFVM(THSystem):
+
+    """This class models 1-D heat diffusion in spherical geometry,
+    and convective heat transfer at the solid surface to fluid.
+    """
+
+    def __init__(self, kappa, components):
+        THSystem.__init__(self, kappa, components)
+
+    def dtempdt(self, component, power, omegas, t_idx):
```

Comment:
This function is a bit long and convoluted. Can it be broken up into smaller functions which you can reuse more efficiently?